File: sdk/python/recursive_sdk/wire.py
"""Parse Claude Code–compatible stream-json NDJSON into SDK message types."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from .models import (
    AssistantMessage,
    PartialAssistantMessage,
    RunResult,
    SystemMessage,
    TextContent,
    ToolResultBlock,
    ToolUseBlock,
    UsageMeta,
    UserMessage,
    _map_finish_reason_to_subtype,
)

WireItem = Dict[str, Any]  # {"kind": "message"|"result"|"session", ...}


def parse_wire_object(
    raw: Dict[str, Any],
    fallback_session_id: str = "",
) -> Optional[WireItem]:
    """Parse one NDJSON object from ``recursive --output-format stream-json``."""
    typ = str(raw.get("type", ""))
    session_id = str(raw.get("session_id") or fallback_session_id)

    if typ == "system":
        subtype = str(raw.get("subtype", ""))
        msg = SystemMessage(type="system", subtype=subtype, data=raw)
        if subtype == "init" and raw.get("session_id"):
            return {
                "kind": "session",
                "session_id": str(raw["session_id"]),
                "message": msg,
            }
        return {"kind": "message", "message": msg}

    if typ == "assistant":
        message = raw.get("message") or {}
        content = _parse_content_blocks(message.get("content"))
        return {
            "kind": "message",
            "message": AssistantMessage(
                type="assistant", content=content, session_id=session_id
            ),
        }

    if typ == "user":
        message = raw.get("message") or {}
        content_raw = message.get("content")
        if isinstance(content_raw, str):
            content = content_raw
        elif isinstance(content_raw, list):
            import json

            content = json.dumps(content_raw)
        else:
            content = str(content_raw or "")
        return {
            "kind": "message",
            "message": UserMessage(
                type="user", content=content, session_id=session_id
            ),
        }

    if typ == "stream_event":
        event = raw.get("event") or {}
        delta = event.get("delta") or {}
        text = str(delta.get("text") or "")
        if not text:
            return None
        return {
            "kind": "message",
            "message": PartialAssistantMessage(
                type="stream_event",
                text=text,
                step=int(event.get("index") or 0),
                session_id=session_id,
            ),
        }

    if typ == "result":
        return {
            "kind": "result",
            "result": _parse_result_object(raw, session_id),
        }

    return None


def _parse_content_blocks(content_raw: Any) -> List[Any]:
    content: List[Any] = []
    if isinstance(content_raw, str):
        content.append(TextContent(text=content_raw))
        return content
    if not isinstance(content_raw, list):
        return content
    for item in content_raw:
        if not isinstance(item, dict):
            continue
        t = item.get("type")
        if t == "text":
            content.append(TextContent(text=str(item.get("text") or "")))
        elif t == "tool_use":
            content.append(
                ToolUseBlock(
                    id=str(item.get("id") or ""),
                    name=str(item.get("name") or ""),
                    input=item.get("input") or {},
                )
            )
        elif t == "tool_result":
            content.append(
                ToolResultBlock(
                    tool_use_id=str(item.get("tool_use_id") or ""),
                    content=str(item.get("content") or ""),
                )
            )
    return content


def _parse_result_object(raw: Dict[str, Any], session_id: str) -> RunResult:
    subtype_raw = str(raw.get("subtype") or "success")
    is_error = bool(raw.get("is_error"))

    if subtype_raw == "error_max_budget_usd":
        subtype = "error_during_execution"
    elif subtype_raw in (
        "success",
        "error_max_turns",
        "error_during_execution",
        "cancelled",
    ):
        subtype = subtype_raw
    else:
        subtype = _map_finish_reason_to_subtype(
            str(raw.get("stop_reason") or ""),
            "error" if is_error else "finished",
        )

    status = "finished"
    if subtype == "cancelled":
        status = "cancelled"
    elif is_error or subtype != "success":
        status = "error"

    usage = None
    usage_raw = raw.get("usage")
    if isinstance(usage_raw, dict):
        usage = UsageMeta(
            input_tokens=int(usage_raw.get("input_tokens") or 0),
            output_tokens=int(usage_raw.get("output_tokens") or 0),
            cache_creation_tokens=(
                int(usage_raw["cache_creation_input_tokens"])
                if usage_raw.get("cache_creation_input_tokens") is not None
                else None
            ),
            cache_read_tokens=(
                int(usage_raw["cache_read_input_tokens"])
                if usage_raw.get("cache_read_input_tokens") is not None
                else None
            ),
        )

    errors = raw.get("errors")
    error = None
    if isinstance(errors, list) and errors:
        error = "; ".join(str(e) for e in errors)
    elif is_error:
        error = subtype_raw

    return RunResult(
        id=session_id,
        status=status,
        finish_reason=_finish_reason_for_subtype(subtype, raw.get("stop_reason")),
        usage=usage,
        error=error,
        result=str(raw["result"]) if raw.get("result") is not None else None,
        num_turns=int(raw["num_turns"]) if raw.get("num_turns") is not None else 0,
        duration_ms=(
            int(raw["duration_ms"]) if raw.get("duration_ms") is not None else None
        ),
    )


def _finish_reason_for_subtype(
    subtype: str,
    stop_reason: Any,
) -> Optional[str]:
    """Pick a finish_reason that makes ``RunResult.subtype`` match the wire label."""
    if stop_reason is not None:
        return str(stop_reason)
    if subtype == "success":
        return "NoMoreToolCalls"
    if subtype == "error_max_turns":
        return "BudgetExceeded"
    if subtype == "cancelled":
        return "Cancelled"
    return subtype

File: sdk/typescript/src/wire.ts
/**
 * Parse Claude Code–compatible stream-json NDJSON into SDK message types.
 */

import { mapFinishReasonToSubtype } from "./models.js";
import type {
  AssistantMessage,
  ContentBlock,
  PartialAssistantMessage,
  RunResult,
  RunSubtype,
  SDKMessage,
  SystemMessage,
  UsageMeta,
  UserMessage,
} from "./models.js";

/** A parsed stream-json line: either a yieldable message or a terminal result. */
export type WireItem =
  | { kind: "message"; message: SDKMessage }
  | { kind: "result"; result: RunResult }
  | { kind: "session"; sessionId: string; message?: SDKMessage };

/**
 * Parse one NDJSON object from `recursive --output-format stream-json`.
 *
 * Returns `null` for lines that should be ignored (malformed / unknown).
 */
export function parseWireObject(
  raw: Record<string, unknown>,
  fallbackSessionId: string,
): WireItem | null {
  const type = String(raw["type"] ?? "");
  const sessionId = String(raw["session_id"] ?? fallbackSessionId);

  if (type === "system") {
    const subtype = String(raw["subtype"] ?? "");
    const msg: SystemMessage = {
      type: "system",
      subtype,
      data: raw,
    };
    if (subtype === "init" && raw["session_id"]) {
      return {
        kind: "session",
        sessionId: String(raw["session_id"]),
        message: msg,
      };
    }
    return { kind: "message", message: msg };
  }

  if (type === "assistant") {
    const message = raw["message"] as Record<string, unknown> | undefined;
    const content = parseContentBlocks(message?.["content"]);
    const msg: AssistantMessage = {
      type: "assistant",
      content,
      sessionId,
    };
    return { kind: "message", message: msg };
  }

  if (type === "user") {
    const message = raw["message"] as Record<string, unknown> | undefined;
    const contentRaw = message?.["content"];
    let content: string;
    if (typeof contentRaw === "string") {
      content = contentRaw;
    } else if (Array.isArray(contentRaw)) {
      content = JSON.stringify(contentRaw);
    } else {
      content = String(contentRaw ?? "");
    }
    const msg: UserMessage = {
      type: "user",
      content,
      sessionId,
    };
    return { kind: "message", message: msg };
  }

  if (type === "stream_event") {
    const event = raw["event"] as Record<string, unknown> | undefined;
    const delta = event?.["delta"] as Record<string, unknown> | undefined;
    const text = String(delta?.["text"] ?? "");
    if (!text) return null;
    const msg: PartialAssistantMessage = {
      type: "stream_event",
      text,
      step: Number(event?.["index"] ?? 0),
      sessionId,
    };
    return { kind: "message", message: msg };
  }

  if (type === "result") {
    return { kind: "result", result: parseResultObject(raw, sessionId) };
  }

  return null;
}

function parseContentBlocks(contentRaw: unknown): ContentBlock[] {
  const content: ContentBlock[] = [];
  if (typeof contentRaw === "string") {
    content.push({ type: "text", text: contentRaw });
    return content;
  }
  if (!Array.isArray(contentRaw)) return content;
  for (const item of contentRaw as Record<string, unknown>[]) {
    const t = item["type"] as string;
    if (t === "text") {
      content.push({ type: "text", text: String(item["text"] ?? "") });
    } else if (t === "tool_use") {
      content.push({
        type: "tool_use",
        id: String(item["id"] ?? ""),
        name: String(item["name"] ?? ""),
        input: (item["input"] as Record<string, unknown>) ?? {},
      });
    } else if (t === "tool_result") {
      content.push({
        type: "tool_result",
        tool_use_id: String(item["tool_use_id"] ?? ""),
        content: String(item["content"] ?? ""),
      });
    }
  }
  return content;
}

function parseResultObject(
  raw: Record<string, unknown>,
  sessionId: string,
): RunResult {
  const subtypeRaw = String(raw["subtype"] ?? "success");
  const isError = Boolean(raw["is_error"]);
  const subtype = normalizeSubtype(subtypeRaw);

  let status: RunResult["status"] = "finished";
  if (subtype === "cancelled") status = "cancelled";
  else if (isError || subtype !== "success") status = "error";

  const usageRaw = raw["usage"] as Record<string, unknown> | undefined;
  const usage: UsageMeta | undefined = usageRaw
    ? {
        inputTokens: Number(usageRaw["input_tokens"] ?? 0),
        outputTokens: Number(usageRaw["output_tokens"] ?? 0),
        cacheCreationTokens:
          usageRaw["cache_creation_input_tokens"] != null
            ? Number(usageRaw["cache_creation_input_tokens"])
            : undefined,
        cacheReadTokens:
          usageRaw["cache_read_input_tokens"] != null
            ? Number(usageRaw["cache_read_input_tokens"])
            : undefined,
      }
    : undefined;

  const errors = raw["errors"] as string[] | undefined;
  const error =
    errors && errors.length > 0
      ? errors.join("; ")
      : isError
        ? subtypeRaw
        : undefined;

  return {
    id: sessionId,
    status,
    subtype:
      subtype === "success" ||
      subtype === "error_max_turns" ||
      subtype === "error_during_execution" ||
      subtype === "cancelled"
        ? subtype
        : mapFinishReasonToSubtype(
            String(raw["stop_reason"] ?? ""),
            status,
          ),
    finishReason: raw["stop_reason"] != null ? String(raw["stop_reason"]) : undefined,
    usage,
    error,
    ok: status === "finished",
    result: raw["result"] != null ? String(raw["result"]) : undefined,
    numTurns:
      raw["num_turns"] != null ? Number(raw["num_turns"]) : undefined,
    durationMs:
      raw["duration_ms"] != null ? Number(raw["duration_ms"]) : undefined,
  };
}

function normalizeSubtype(raw: string): RunSubtype | string {
  if (
    raw === "success" ||
    raw === "error_max_turns" ||
    raw === "error_during_execution" ||
    raw === "cancelled" ||
    raw === "error_max_budget_usd"
  ) {
    if (raw === "error_max_budget_usd") return "error_during_execution";
    return raw;
  }
  return raw;
}
